test/e2e/provisioner/provisioner_test.go

```diff
@@ -489,6 +489,176 @@ var _ = Describe("Gateway provisioner", func() {
+			})
+			require.True(f.T(), ok, fmt.Sprintf("gateway is %v", gateway))
+			type testObj struct {
+				expectReconcile bool
+				namespace       string
+			}
+			testcases := []testObj{
+				{
+					expectReconcile: true,
+					namespace:       "testns-1",
+				},
+				{
+					expectReconcile: true,
+					namespace:       "testns-2",
+				},
+				{
+					expectReconcile: false,
+					namespace:       "testns-3",
+				},
+			}
+
+			By("Deploy workload in target namespaces, check if they get reconciled or not")
+			for _, t := range testcases {
+				f.Fixtures.Echo.Deploy(t.namespace, "echo")
+
+				route := &gatewayapi_v1beta1.HTTPRoute{
+					ObjectMeta: metav1.ObjectMeta{
+						Namespace: t.namespace,
+						Name:      "httproute-1",
+					},
+					Spec: gatewayapi_v1beta1.HTTPRouteSpec{
+						Hostnames: []gatewayapi_v1beta1.Hostname{"provisioner.projectcontour.io"},
+						CommonRouteSpec: gatewayapi_v1beta1.CommonRouteSpec{
+							ParentRefs: []gatewayapi_v1beta1.ParentReference{
+								gatewayapi.GatewayParentRef("", gateway.Name),
+							},
+						},
+						Rules: []gatewayapi_v1beta1.HTTPRouteRule{
+							{
+								Matches:     gatewayapi.HTTPRouteMatch(gatewayapi_v1.PathMatchPathPrefix, "/prefix"),
+								BackendRefs: gatewayapi.HTTPBackendRef("echo", 80, 1),
+							},
+						},
+					},
+				}
+
+				if t.expectReconcile {
+					// set route's parentRef's namespace to the gateway's namespace
+					route.Spec.CommonRouteSpec.ParentRefs[0].Namespace = (*gatewayapi_v1.Namespace)(&namespace)
+					// set the route's hostnames to custom name with namespace inside
+					route.Spec.Hostnames = []gatewayapi_v1beta1.Hostname{gatewayapi_v1beta1.Hostname("provisioner.projectcontour.io." + t.namespace)}
+
+					By(fmt.Sprintf("Expect namespace %s to be watched by contour", t.namespace))
+					hr, ok := f.CreateHTTPRouteAndWaitFor(route, e2e.HTTPRouteAccepted)
+					By(fmt.Sprintf("Expect httproute under namespace %s is accepted", t.namespace))
+					require.True(f.T(), ok, fmt.Sprintf("httproute is %v", hr))
+					res, ok := f.HTTP.RequestUntil(&e2e.HTTPRequestOpts{
+						OverrideURL: "http://" + net.JoinHostPort(gateway.Status.Addresses[0].Value, "80"),
+						Host:        string(route.Spec.Hostnames[0]),
+						Path:        "/prefix/match",
+						Condition:   e2e.HasStatusCode(200),
+					})
+					require.NotNil(f.T(), res)
+					require.Truef(f.T(), ok, "expected 200 response code, got %d", res.StatusCode)
+
+					body := f.GetEchoResponseBody(res.Body)
+					assert.Equal(f.T(), t.namespace, body.Namespace)
+					assert.Equal(f.T(), "echo", body.Service)
+				} else {
+					// Root proxy in non-watched namespace should fail
+					By(fmt.Sprintf("Expect namespace %s not to be watched by contour", t.namespace))
+					hr, ok := f.CreateHTTPRouteAndWaitFor(route, e2e.HTTPRouteIngnoredByContour)
+					By(fmt.Sprintf("Expect httproute under namespace %s is not accepted", t.namespace))
```

Comment:
nit: we could remove this line and the below since it is covered by the `require.Never`